### PR TITLE
Add `Projects` router.

### PR DESCRIPTION
**Pull request by FledgeXu**
## Goal
The purpose of this PR is to add the `projects` endpoint to router.

## Brief
This PR includes three main changes:
1. Change the `collections` table to `projects` table to reflect the changing of the UI.
2. Add Setting Cookie to `users` router and use the right plural endpoint name.
3. Add the `projects` router.

### CODE REVIEW

File: backend/tests/test_routers.py [hunk 1]
```diff
@@ -1,18 +1,95 @@
...
+    assert uuid.UUID(json_result.get("id"))
+    assert parser.parse(json_result.get("created_on"))
+    assert response.headers.get("set-cookie") is not None
+
+
+def test_user_validation():
+    response = client.post(f"{API_VERSION_PREFIX}/users/projects")
+    assert response.status_code == httpx.codes.UNAUTHORIZED
+    params = {"user_id": NON_EXISTENT_UUID}
+    response = client.post(f"{API_VERSION_PREFIX}/users/projects", params=params)
```
Comment (rgaudin) [inline backend/tests/test_routers.py:29]: We should use `cookie`

File: backend/tests/test_routers.py [hunk 2]
```diff
@@ -1,18 +1,95 @@
...
     assert response.status_code == httpx.codes.CREATED
     json_result = response.json()
     assert uuid.UUID(json_result.get("id"))
+    assert uuid.UUID(json_result.get("user_id"))
+    assert json_result.get("name") is not None
+    assert parser.parse(json_result.get("created_on"))
+    assert parser.parse(json_result.get("expire_on"))
+
+
+@pytest.fixture
```
Comment (rgaudin) [inline backend/tests/test_routers.py:52]: Move it ; makes tests less readable

File: backend/tests/test_routers.py [hunk 3]
```diff
@@ -1,18 +1,95 @@
...
+def create_project(create_user):
+    user_id = create_user
+    params = {"name": "project1", "user_id": user_id}
+    response = client.post(f"{API_VERSION_PREFIX}/users/projects", params=params)
+    json_result = response.json()
+    return (json_result.get("id"), create_user)
```
Comment (rgaudin) [inline backend/tests/test_routers.py:58]: fixture name doesn't describe the returned object and the returned object is a tuple. This is hardly readable/usable

File: backend/tests/test_routers.py [hunk 4]
```diff
@@ -1,18 +1,95 @@
...
+
+
+def test_get_all_projects(create_project):
+    params = {"user_id": create_project[1]}
```
Comment (rgaudin) [inline backend/tests/test_routers.py:62]: See ? We don't want people to wonder what `create_project[1]` should be

File: backend/tests/test_routers.py [hunk 5]
```diff
@@ -1,18 +1,95 @@
...
+    response = client.get(f"{API_VERSION_PREFIX}/users/projects", params=params)
+    json_result = response.json()
+    assert json_result is not None
+    assert len(json_result) == 1
```
Comment (rgaudin) [inline backend/tests/test_routers.py:66]: You can also check that known project ID is inside. Otherwise list schema is not tested

File: backend/tests/test_routers.py [hunk 6]
```diff
@@ -1,18 +1,95 @@
...
+
+
+def test_get_project(create_project):
+    params = {"id": create_project[0], "user_id": create_project[1]}
+    response = client.get(f"{API_VERSION_PREFIX}/users/projects/", params=params)
+    json_result = response.json()
+    assert uuid.UUID(json_result.get("id"))
+    assert uuid.UUID(json_result.get("user_id"))
+    assert json_result.get("name") is not None
```
Comment (rgaudin) [inline backend/tests/test_routers.py:75]: You should chain fixture so you have a `project_name` that feeds a `project_id` (which creates it) and then here you can use both fixture and test that the result does have the fixture name

File: backend/tests/test_routers.py [hunk 7]
```diff
@@ -1,18 +1,95 @@
...
+def test_get_all_projects(create_project):
+    params = {"user_id": create_project[1]}
+    response = client.get(f"{API_VERSION_PREFIX}/users/projects", params=params)
+    json_result = response.json()
+    assert json_result is not None
+    assert len(json_result) == 1
+
+
+def test_get_project(create_project):
+    params = {"id": create_project[0], "user_id": create_project[1]}
```
Comment (rgaudin) [inline backend/tests/test_routers.py:70]: we should not have a single generic endpoint and passing it those IDs.

It should be a qualified endpoint `/projects/{project_id}` that gets the `user_id` from/as the cookie.

File: backend/tests/test_routers.py [hunk 8]
```diff
@@ -1,18 +1,95 @@
...
+    assert response.status_code == httpx.codes.NOT_FOUND
+
+
+def test_update_project(create_project):
+    params = {
+        "name": "updated_name",
+        "id": create_project[0],
+        "user_id": create_project[1],
+    }
+    response = client.put(f"{API_VERSION_PREFIX}/users/projects", params=params)
```
Comment (rgaudin) [inline backend/tests/test_routers.py:90]: You don't check the status code

File: backend/tests/test_routers.py [hunk 9]
```diff
@@ -1,18 +1,95 @@
...
+
+    params = {"id": create_project[0], "user_id": create_project[1]}
+    response = client.get(f"{API_VERSION_PREFIX}/users/projects/", params=params)
+    json_result = response.json()
+    assert json_result.get("name") == "updated_name"
```
Comment (rgaudin) [inline backend/tests/test_routers.py:95]: We should test all endpoints for unauthorized access: no ID or incorrect one

**Review by rgaudin — CHANGES_REQUESTED** [review]

File: backend/src/backend/routes/projects.py [hunk 10]
```diff
@@ -8,94 +8,93 @@
...
-    statement = select(Project).filter_by(user_id=user.id).filter_by(id=id)
-    queried_project = session.execute(statement).scalar()
-    if not queried_project:
-        raise HTTPException(codes.NOT_FOUND, "Not found project.")
-    return ProjectModal.from_orm(queried_project)
+    return ProjectModel.from_orm(project)
 
 
-@router.put("")
+@router.put("/{project_id}")
```
Comment (rgaudin) [inline backend/src/backend/routes/projects.py:92]: This should be a PATCH request as we're just updating some details and not replacing.

File: backend/tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@pytest.fixture()
+def test_client():
+    client = TestClient(app)
+    return client
+
+
+@pytest.fixture
+def non_existent_uuid():
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/tests/conftest.py:15]: Ah I see it's used for project as well. Not a good idea as changing it could have unexpected consequences elsewhere. Use two fixtures, it's cheap 😉

**Review by rgaudin — CHANGES_REQUESTED** [review]
OK it's taking shape.

Tests should be refactored though:
- `tests/routes/test_users.py` instead of `tests/test_users_route.py` (same for project)
- we'd need Models (DB) tests as well at some point. This can wait (as it's still moving) so maybe open a ticket and we'll tackle that once we have all the routes.
- fixtures and tests that create data should remove it as well so tests are reproducible. fixtures usually yield the generated data and delete it during the test teardown (after the `yield` statement). For tests that generate data, one approach (see zimfarm) is to declare this data and a garbage-collector fixture takes care of removing it.
- tests should be focused on one scope and named accordingly: one `test_get_project` and one `test_get_project_incorrect_id` for instance.
- validations should be on separate tests for the different endpoints so endpoints are _grouped_ and can be easily removed/commented/excluded.

File: backend/src/backend/database/models.py [hunk 12]
```diff
@@ -99,7 +99,7 @@ class File(Base):
     id: Mapped[UUID] = mapped_column(
         init=False, primary_key=True, server_default=text("uuid_generate_v4()")
     )
-    project_id: Mapped[UUID] = mapped_column(ForeignKey("projects.id"), init=False)
+    projects_id: Mapped[UUID] = mapped_column(ForeignKey("projects.id"), init=False)
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/src/backend/database/models.py:102]: In the last review, you told me it's a typo😂. I will change it then.

File: backend/tests/conftest.py [hunk 13]
```diff
@@ -34,8 +39,14 @@ def logged_in_client(test_client) -> str:
...
     data = {"name": test_project_name}
     response = logged_in_client.post(f"{API_VERSION_PREFIX}/projects", json=data)
     json_result = response.json()
     return json_result.get("id")
+
+
+@pytest.fixture()
+def test_project_id(logged_in_client, created_project_id):
+    yield created_project_id
+    logged_in_client.delete(f"{API_VERSION_PREFIX}/projects/{created_project_id}")
```
Comment (rgaudin) [inline backend/tests/conftest.py:52]: Good ; you don't need to use the API to create/delete your fixtures. It makes testing your API more difficult anyway

File: backend/src/backend/routes/projects.py [hunk 14]
```diff
@@ -89,12 +90,20 @@ async def get_project(project=Depends(validated_project)) -> ProjectModel:
     return ProjectModel.from_orm(project)
 
 
-@router.put("/{project_id}")
+@router.delete("/{project_id}")
+async def delete_project(
+    project=Depends(validated_project), session=Depends(gen_session)
+):
+    """Delete a specific project by its id."""
+    session.delete(project)
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/src/backend/routes/projects.py:98]: And if you don't return anything, use `204`

File: backend/tests/routes/test_projects.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    )
+
+    assert response.status_code == codes.OK
+
+    response = logged_in_client.get(f"{API_VERSION_PREFIX}/projects/{test_project_id}")
+    json_result = response.json()
+    assert json_result.get("name") == "updated_name"
+
+
+def test_update_project_wrong_data(logged_in_client, non_existent_project_uuid):
```
Comment (rgaudin) [inline backend/tests/routes/test_projects.py:121]: It's not wrong_data but wrong_id

**Review by rgaudin — CHANGES_REQUESTED** [review]
Great! Still a couple of things to fix.

File: backend/src/backend/database/models.py [hunk 16]
```diff
@@ -99,7 +99,7 @@ class File(Base):
     id: Mapped[UUID] = mapped_column(
         init=False, primary_key=True, server_default=text("uuid_generate_v4()")
     )
-    projects_id: Mapped[UUID] = mapped_column(ForeignKey("projects.id"), init=False)
+    project_id: Mapped[UUID] = mapped_column(ForeignKey("projects.id"), init=False)
```
[the remark this answers is not in the document]
Comment (FledgeXu) [reply backend/src/backend/database/models.py:102]: I think this `s` is necessary, cuz the table name is `Porjects` not `Project`.

[the remark this answers is not in the document]
Comment (rgaudin) [reply backend/src/backend/database/models.py:102]: Oh you're right ; my bad.
We should rename models and tables to singular though. That's the convenient convention.
Can be in a separate PR though

File: backend/tests/conftest.py (hunk 13 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:52]: > Good ; you don't need to use the API to create/delete your fixtures. It makes testing your API more difficult anyway

I don't get it here. If I don't use the API to create/delete project, How I can create/delete the project?

Comment (rgaudin) [reply backend/tests/conftest.py:52]: Those tests are integration tests. They test endpoints behavior.
If you want to test endpoints independently (that's what we do we our per-endpoint tests), you may not want to depend on another endpoint behaving correctly. Otherwise, if that dependency endpoint fails, then all your tests fail and you have to fix that first because without it fixed, your tests are useless.
To avoid this, you can use the DB to create your fixtures (insert and delete data) ; assuming you already, independently test all endpoints (and eventually end-to-end test)
But all approaches are valid; just wanted to make sure you knew your options.

File: backend/tests/conftest.py [hunk 17]
```diff
@@ -1,18 +1,33 @@
...
+from backend.database import Session
+from backend.database.models import Project, User
 from src.backend.entrypoint import app
 
 
 @pytest.fixture()
-def test_client():
+def mock_user_id():
+    new_user = User(created_on=datetime.utcnow(), projects=[])
+    with scoped_session(Session)() as session:
```
Comment (rgaudin) [inline backend/tests/conftest.py:15]: Can you elaborate on the need for `scoped_session`? From [the doc](https://docs.sqlalchemy.org/en/20/orm/contextual.html#unitofwork-contextual) I fail to see why we'd need it.
Wouldn't a transaction inside the session be sufficient?

I see that this created User will never be removed from the DB.
I suggested previously to look at [zimfarm's garbage collector](https://github.com/openzim/zimfarm/blob/main/dispatcher/backend/src/tests/integration/conftest.py#L84).

I think the approach is better: you declare what you created and it takes care of deleting it.

File: backend/tests/conftest.py [hunk 18]
```diff
@@ -1,18 +1,33 @@
...
+        session.add(new_user)
+        session.flush()
+        session.refresh(new_user)
+        session.commit()
+        return new_user.id
+
+
+@pytest.fixture()
+def mock_client():
```
Comment (rgaudin) [inline backend/tests/conftest.py:24]: I think the `mock_` prefix is superfluous here. Also, why not directly returning? The variable is not necessary.

I also think this fixture should be session-scoped (as some others ; check the [doc](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html?highlight=scope#scope-sharing-fixtures-across-classes-modules-packages-or-session))

File: backend/tests/conftest.py [hunk 19]
```diff
@@ -32,21 +47,33 @@ def test_project_name():
...
+    return mock_client
 
 
 @pytest.fixture()
-def created_project_id(logged_in_client, test_project_name):
-    data = {"name": test_project_name}
-    response = logged_in_client.post(f"{API_VERSION_PREFIX}/projects", json=data)
-    json_result = response.json()
-    return json_result.get("id")
+def created_project_id(test_project_name, mock_user_id):
```
Comment (rgaudin) [inline backend/tests/conftest.py:57]: `project_id` seems fine as this fixture's name. Use prefix to differentiate. here, it's the normal, expected, behavior.
Mind the scope as well.

File: backend/tests/conftest.py [hunk 20]
```diff
@@ -32,21 +47,33 @@ def test_project_name():
...
+        session.add(new_project)
+        session.flush()
+        session.refresh(new_project)
+        session.commit()
+        return new_project.id
 
 
 @pytest.fixture()
-def test_project_id(logged_in_client, created_project_id):
+def mock_project_id(created_project_id):
```
Comment (rgaudin) [inline backend/tests/conftest.py:74]: This one looks like a hack. Can't you yield from the above function and delete afterward?

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you ; I have some questions regarding the cleanup approach ; see inline comments.

File: backend/tests/conftest.py (hunk 18 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:24]: I don't think we should use `scope="session"` here. Because that will bring the side effects to the tests (e.g. legacy cookies). Overall, it's a `TestClient` and it's pretty lightweight.

File: backend/tests/conftest.py (hunk 17 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:15]: The the reason why I use `scoped_session` is that I found if I don't use it when I call the `session.begin()` it will complain the thread error. Let me try again.

Comment (FledgeXu) [reply backend/tests/conftest.py:15]: I know why. I added an extra `commit` by mistake.

File: backend/tests/conftest.py (hunk 18 above)
Comment (rgaudin) [reply backend/tests/conftest.py:24]: OK

File: backend/tests/conftest.py (hunk 20 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:74]: Actually I don't think we need to delete the project anyway. we can rename `created_project_id` to `project_id` and delete the `mock_project_id` whole function.

File: backend/tests/conftest.py (hunk 19 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:57]: The `project_id` should also be in the `function` scope. Each test should have its own project to avoid side effects.

File: backend/tests/conftest.py (hunk 20 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:74]: The advantage of doing so is that we can check the database to see the original data.

File: backend/tests/conftest.py [hunk 21]
```diff
@@ -1,29 +1,26 @@
...
 from backend.database import Session
 from backend.database.models import Project, User
 from src.backend.entrypoint import app
 
 
 @pytest.fixture()
-def mock_user_id():
-    new_user = User(created_on=datetime.utcnow(), projects=[])
-    with scoped_session(Session)() as session:
+def user_id():
```
Comment (rgaudin) [inline backend/tests/conftest.py:12]: OK but now we're not deleting at all.
Here's how to:

```py
@pytest.fixture()
def user_id():
    new_user = User(created_on=datetime.datetime.utcnow(), projects=[])
    with Session.begin() as session:
        session.add(new_user)
        session.flush()
        session.refresh(new_user)
        created_id = new_user.id
    yield created_id
    with Session.begin() as session:
        user = session.get(User, created_id)
        if user:
            session.delete(user)

...

@pytest.fixture()
def project_id(test_project_name, user_id):
    now = datetime.datetime.utcnow()
    new_project = Project(
        name=test_project_name, created_on=now, expire_on=None, files=[], archives=[]
    )
    with Session.begin() as session:
        user = session.get(User, user_id)
        if user:
            user.projects.append(new_project)
        session.add(new_project)
        session.flush()
        session.refresh(new_project)
        created_id = new_project.id
    yield created_id
    with Session.begin() as session:
        project = session.get(User, user_id)
        if project:
            session.delete(project)
```

**Review by rgaudin — CHANGES_REQUESTED** [review]

File: backend/tests/conftest.py (hunk 21 above)
Comment (FledgeXu) [reply backend/tests/conftest.py:12]: Ok

**Review by rgaudin — APPROVED** [review]
👍
